File: mobile/src/main/java/com/tilly/steven/stlbusarrivals/ui/fragments/MapsFragment.java
package com.tilly.steven.stlbusarrivals.ui.fragments;

import android.Manifest;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Color;
import android.os.Bundle;
import android.os.Handler;
import android.support.v4.app.ActivityCompat;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import com.android.volley.RequestQueue;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.StringRequest;
import com.tilly.steven.stlbusarrivals.Model.PathBounds;
import com.tilly.steven.stlbusarrivals.Model.PathList;
import com.tilly.steven.stlbusarrivals.Model.Point;
import com.tilly.steven.stlbusarrivals.Model.VehiculeList;
import com.tilly.steven.stlbusarrivals.R;
import com.tilly.steven.stlbusarrivals.VolleySingleton;
import com.tilly.steven.stlbusarrivals.XmlParser;
import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.SupportMapFragment;
import com.google.android.gms.maps.model.BitmapDescriptorFactory;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.LatLngBounds;
import com.google.android.gms.maps.model.MarkerOptions;
import com.google.android.gms.maps.model.PolylineOptions;

import java.util.ArrayList;
import java.util.Observable;
import java.util.Observer;

/**
 * Created by Steven on 2016-01-28.
 */
public class MapsFragment extends Fragment implements OnMapReadyCallback, Observer {

    public MapsFragment() {
    }

    private GoogleMap mMap;
    private ArrayList<ArrayList<Point>> pathList;
    private VehiculeList vehiculeList;
    private XmlParser xmlparser = new XmlParser();
    private String longitude, latitude, routeTag;
    private PathBounds pathBounds;
    private Handler mHandler;


    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        xmlparser.addObserver(this);
        routeTag = getArguments().getString("routeTag");
        mHandler = new Handler();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        ViewGroup v = (ViewGroup) inflater.inflate(R.layout.fragment_map, container, false);
        // Obtain the SupportMapFragment and get notified when the map is ready to be used.
        SupportMapFragment mapFragment = (SupportMapFragment) getChildFragmentManager().
                findFragmentById(R.id.map);
        mapFragment.getMapAsync(this);
        return v;
    }

    @Override
    public void onResume() {
        super.onResume();
    }

    void startRepeatingTask() {
        mStatusChecker.run();
    }

    @Override
    public void onStop() {
        super.onStop();
        mHandler.removeCallbacks(mStatusChecker);
    }

    Runnable mStatusChecker = new Runnable() {
        @Override
        public void run() {
            sendDetailRequest();
            int mInterval = 10000;
            mHandler.postDelayed(mStatusChecker, mInterval);
        }
    };

    private void sendPathRequest() {
        RequestQueue queue = VolleySingleton.getInstance(getActivity()).getRequestQueue();
        xmlparser.addObserver(this);
        String url = "http://webservices.nextbus.com/service/publicXMLFeed?command=routeConfig&a=stl&r=" + routeTag;
        StringRequest request = new StringRequest(url, new Response.Listener<String>() {

            @Override
            public void onResponse(String response) {
                // we got the response, now our job is to handle it
                //parseXmlResponse(response);
                try {
                    xmlparser.readStopXml(response);
                } catch(Exception e) {
                    e.printStackTrace();
                }

            }
        }, new Response.ErrorListener() {

            @Override
            public void onErrorResponse(VolleyError error) {
                error.printStackTrace();
                sendPathRequest();
            }
        });
        queue.add(request);
    }

    private void sendDetailRequest() {
        RequestQueue queue = VolleySingleton.getInstance(getActivity()).getRequestQueue();
        String url = "http://webservices.nextbus.com/service/publicXMLFeed?command=vehicleLocations&a=stl&r=" + routeTag + "&t=0";
        StringRequest request = new StringRequest(url, new Response.Listener<String>() {

            @Override
            public void onResponse(String response) {
                // we got the response, now our job is to handle it
                //parseXmlResponse(response);
                try {
                    xmlparser.readLocation(response);
                } catch(Exception e) {
                    e.printStackTrace();
                }

            }
        }, new Response.ErrorListener() {

            @Override
            public void onErrorResponse(VolleyError error) {
                error.printStackTrace();
                sendDetailRequest();
            }
        });
        queue.add(request);
    }

    @Override
    public void update(Observable observable, Object o) {
        if(getActivity() != null) {
            if(o instanceof VehiculeList) {
                vehiculeList = (VehiculeList) o;
                if(vehiculeList.size() != 0) {
                    mMap.clear();
                    for(int i = 0; i < vehiculeList.size(); i++) {
                        longitude = vehiculeList.get(i).getLongitude();
                        latitude = vehiculeList.get(i).getLatitude();
                        if(!longitude.equals("false")) {
                            LatLng bus = new LatLng(Double.valueOf(latitude), Double.valueOf(longitude));
                            mMap.addMarker(new MarkerOptions().position(bus).title(getString(R.string.bus_location))
                                    .icon(BitmapDescriptorFactory.fromBitmap(mapIcon()))
                                    .anchor(0.5f, 0.5f));
                        }
                    }
                } else {
                    mMap.clear();
                    Toast.makeText(getActivity(), getString(R.string.server_no_bus), Toast.LENGTH_SHORT).show();
                }
                for(int j = pathList.size() - 1; j >= 0; j--) {
                    for(int i = pathList.get(j).size() - 1; i > 0; i--) {

                        mMap.addPolyline(new PolylineOptions()
                                .add(pathList.get(j).get(i).getLatLng(), pathList.get(j).get(i - 1).getLatLng())
                                .width(5)
                                .color(Color.BLUE));
                    }
                }
            }
            if(o instanceof PathList) {
                pathList = (ArrayList<ArrayList<Point>>) o;
                startRepeatingTask();
            }
            if(o instanceof PathBounds) {
                pathBounds = (PathBounds) o;
                LatLngBounds bounds = new LatLngBounds(
                        pathBounds.getMinBounds(), pathBounds.getMaxBounds());
                mMap.moveCamera(CameraUpdateFactory.newLatLngBounds(bounds, 15));
            }
        }
    }

    private Bitmap mapIcon() {
        Bitmap imageBitmap = BitmapFactory.decodeResource(getResources(), R.mipmap.map_marker);
        return Bitmap.createScaledBitmap(imageBitmap, 64, 64, false);
    }

    /**
     * Manipulates the map once available.
     * This callback is triggered when the map is ready to be used.
     * This is where we can add markers or lines, add listeners or move the camera. In this case,
     * we just add a marker near Sydney, Australia.
     * If Google Play services is not installed on the device, the user will be prompted to install
     * it inside the SupportMapFragment. This method will only be triggered once the user has
     * installed Google Play services and returned to the app.
     */
    @Override
    public void onMapReady(GoogleMap googleMap) {
        mMap = googleMap;
        if(ActivityCompat.checkSelfPermission(getActivity(), Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(getActivity(), Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            // TODO: Consider calling
            //    ActivityCompat#requestPermissions
            // here to request the missing permissions, and then overriding
            //   public void onRequestPermissionsResult(int requestCode, String[] permissions,
            //                                          int[] grantResults)
            // to handle the case where the user grants the permission. See the documentation
            // for ActivityCompat#requestPermissions for more details.
            return;
        }
        mMap.setMyLocationEnabled(true);
        sendPathRequest();

    }
}

File: mobile/src/main/java/com/tilly/steven/stlbusarrivals/ui/fragments/MapsFragment.kt
package com.tilly.steven.stlbusarrivals.ui.fragments

import android.Manifest
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Color
import android.os.Bundle
import android.os.Handler
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.SupportMapFragment
import com.google.android.gms.maps.model.*
import com.tilly.steven.stlbusarrivals.R
import com.tilly.steven.stlbusarrivals.Utils
import com.tilly.steven.stlbusarrivals.Utils.toast
import com.tilly.steven.stlbusarrivals.model.PathBounds
import com.tilly.steven.stlbusarrivals.model.PathList
import com.tilly.steven.stlbusarrivals.model.VehiculeList
import com.tilly.steven.stlbusarrivals.network.DetailsRepo
import com.tilly.steven.stlbusarrivals.network.NetworkCallback

/**
 * Created by Steven on 2016-01-28.
 */
class MapsFragment : Fragment(), OnMapReadyCallback, NetworkCallback {

    private var mMap: GoogleMap? = null
    private var pathList: PathList = PathList()
    private var routeTag: String = ""
    private lateinit var mHandler: Handler

    private var mStatusChecker: Runnable = object : Runnable {
        override fun run() {
            DetailsRepo.getInstance().getLocation(routeTag, this@MapsFragment)
            val mInterval = 10000
            mHandler.postDelayed(this, mInterval.toLong())
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        routeTag = arguments?.getString("routeTag") ?: ""
        mHandler = Handler()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val v = inflater.inflate(R.layout.fragment_map, container, false) as ViewGroup
        // Obtain the SupportMapFragment and get notified when the map is ready to be used.
        val mapFragment = childFragmentManager.findFragmentById(R.id.map) as SupportMapFragment?
        mapFragment?.getMapAsync(this)
        return v
    }

    private fun startRepeatingTask() {
        mStatusChecker.run()
    }

    override fun onStop() {
        super.onStop()
        mHandler.removeCallbacks(mStatusChecker)
    }

    override fun onApiError(error: String) {
        toast(error)
    }

    override fun onPathListLoaded(pathList: PathList) {
        this.pathList = pathList
        startRepeatingTask()
    }

    override fun onVehicleListLoaded(vehiculeList: VehiculeList) {
        if (vehiculeList.size != 0) {
            mMap!!.clear()
            for (i in vehiculeList.indices) {
                val longitude = vehiculeList[i].longitude
                val latitude = vehiculeList[i].latitude
                if (longitude != "false") {
                    val bus = LatLng(java.lang.Double.valueOf(latitude), java.lang.Double.valueOf(longitude))
                    mMap!!.addMarker(MarkerOptions().position(bus).title(getString(R.string.bus_location))
                            .icon(BitmapDescriptorFactory.fromBitmap(mapIcon()))
                            .anchor(0.5f, 0.5f))
                }
            }
        } else {
            mMap!!.clear()
            Toast.makeText(activity, getString(R.string.server_no_bus), Toast.LENGTH_SHORT).show()
        }
        for (j in pathList.indices.reversed()) {
            for (i in pathList[j].size - 1 downTo 1) {

                mMap!!.addPolyline(PolylineOptions()
                        .add(pathList[j][i].getLatLng(), pathList[j][i - 1].getLatLng())
                        .width(5f)
                        .color(Color.BLUE))
            }
        }
    }

    override fun onPathBoundsLoaded(pathBounds: PathBounds) {
        val bounds = LatLngBounds(
                pathBounds.getMinBounds(), pathBounds.getMaxBounds())
        mMap?.moveCamera(CameraUpdateFactory.newLatLngBounds(bounds, 15))
    }

    private fun mapIcon(): Bitmap {
        val imageBitmap = BitmapFactory.decodeResource(resources, R.mipmap.map_marker)
        return Bitmap.createScaledBitmap(imageBitmap, 64, 64, false)
    }

    /**
     * Manipulates the map once available.
     * This callback is triggered when the map is ready to be used.
     * This is where we can add markers or lines, add listeners or move the camera. In this case,
     * we just add a marker near Sydney, Australia.
     * If Google Play services is not installed on the device, the user will be prompted to install
     * it inside the SupportMapFragment. This method will only be triggered once the user has
     * installed Google Play services and returned to the app.
     */
    override fun onMapReady(googleMap: GoogleMap) {
        mMap = googleMap
        if (ActivityCompat.checkSelfPermission(activity!!, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(activity!!, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            // TODO: Consider calling

            val permissionCheck = ContextCompat.checkSelfPermission(
                    context!!, Manifest.permission.ACCESS_FINE_LOCATION)
            if (permissionCheck != PackageManager.PERMISSION_GRANTED) {
                if (ActivityCompat.shouldShowRequestPermissionRationale(activity!!,
                                Manifest.permission.ACCESS_FINE_LOCATION)) {
                    Utils.toast( "Permission Needed")
                } else {
                    requestPermissions(arrayOf(Manifest.permission.ACCESS_FINE_LOCATION), 100)
                }
            } else {
                Toast.makeText(context, "Permission (already) Granted!", Toast.LENGTH_SHORT).show()
            }
            return
        }
        mMap?.isMyLocationEnabled = true
        DetailsRepo.getInstance().getPath(routeTag, this)

    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>,
                                            grantResults: IntArray) {
        when (requestCode) {
            100 -> if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                if (ActivityCompat.checkSelfPermission(context!!, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(context!!, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
                    return
                }
                mMap!!.isMyLocationEnabled = true
                DetailsRepo.getInstance().getPath(routeTag, this)
                Toast.makeText(context, "Permission Granted!", Toast.LENGTH_SHORT).show()
            } else {
                Toast.makeText(context, "Permission Denied!", Toast.LENGTH_SHORT).show()
            }
        }
    }
}
